README.md

```diff
@@ -1,2 +1,98 @@
+  the config above. The proxy serves HTTPS requests at port 8080.
+
+## Configuration
+
+Configuration is typically done with a yaml file. Refer to [config_example.yaml](./example/config_example.yaml) for all 
+configurable options.
+
+### Configuration override
+
+One can override any `string` field (list field override is not supported) with `--set` flag or environment variable.
+
+#### Override config using `--set` flag
+
+To override specific fields, `--set key=value` flag can be set multiple times.
+For example, `proxy.signer.keyId` and `log.level` in the yaml file above can be overridden by:
+
+```shell
+./signing-proxy --config <config_file_path> \
+  --set proxy.signer.keyid=5099392e-3040-40f9-ac70-ce66a9ee0ed6 \
+  --set log.level=debug
+```
+
+#### Override config using env var
+
+A `a.b.c` field can be automatically overridden by setting a `A_B_C` env var
+(all capitalised and dots replaced by underscore).
+For example, `proxy.signer.keyId` and `proxy.signer.bodyDigestAlgo` in the yaml file above can be overridden by:
+
+```shell
+export PROXY_SIGNER_KEYID=5099392e-3040-40f9-ac70-ce66a9ee0ed6
+export PROXY_SIGNER_BODYDIGESTALGO=SHA-512
+```
+
+## Proxy mechanism
+
+Any request coming in the proxy will be signed and forwarded to the upstream target, meaning the host will be replaced
+by the target host and a signature header will be added, the rest of the request is kept as-is. If the request fails 
+the signing validation (due to missing headers for example), the request will not be proxied and the server will return 
+a `400 - Bad Request` response to the client.
+
+The upstream target can be another proxy. In that case, `HTTP_PROXY`, `HTTPS_PROXY` and `NO_PROXY` environment variables 
+can be explicitly set.
+
+Additionally, there are two endpoints explicitly exposed by the proxy:
+
+- `GET /-/health` for health check.
+- `GET /-/prometheus` for metrics.
+
+Incoming requests like above will not be signed nor forwarded to the upstream target.
+
+## Metrics
+
+The proxy publishes certain metrics under `GET /-/prometheus` endpoint:
+
+|       Metric name        |   Type    | Description                                                                        |
+|:------------------------:|:---------:|------------------------------------------------------------------------------------|
+|   internal_error_total   |  Counter  | Total number of the proxy's internal errors. Upstream errors do not count.         |
+|   request_count_total    |  Counter  | Total number of requests coming to the proxy.                                      |
+|   signed_request_total   |  Counter  | Total number of incoming requests that have been signed and proxied.               |
+| signing_duration_seconds | Histogram | Request signing duration time in seconds.                                          |
+| request_duration_seconds | Histogram | Total request duration time in seconds, including signing and upstream processing. |
```

Comment:
I feel that we should keep good documentation standards for this OSS project and provide instructions for testing, development and contributing in the README file.